### ISSUE TITLE
Problems with releasing RAM.

### ISSUE THREAD

**Issue by SiiwY** [0x1mason/GribApi.NET#106]
RAM memory is not released. I create REST Api and a large number of requests hangs the application.
<br/>
![image](https://user-images.githubusercontent.com/24372059/50085114-6be27b00-01f9-11e9-907c-e5953ac34a13.png)
<br/>
![image](https://user-images.githubusercontent.com/24372059/50085136-79980080-01f9-11e9-8850-dcb3f60f2787.png)
<br/>

![image](https://user-images.githubusercontent.com/24372059/50085451-5faaed80-01fa-11e9-8764-4416428d9928.png)
<br/>
```c#
private  GeoSpatialValue GetWave(string path)
{
    GeoSpatialValue value;
    using (GribFile file = new GribFile(path))
    {
        var msg = file.Where(d => d.StartStep == timeFile.Hours.ToString()).FirstOrDefault();
        value = msg.GeoSpatialValues.Where(p => p.Latitude == parameters.Latitude).Where(p => p.Longitude == parameters.Longitude).FirstOrDefault();
    }
    return value;
}
```
<br/>
It seems to me that this is a mistake in your application.
Can I help you?

**Comment by jnyrup** [comment]
Please try to upgrade to the newest 1.0.0-beta4, to see if that resolves your problem.
Note that `GeoSpatialValue` is deprecated in the new version, in favor of `GridCoordinateValue`.

**Comment by SiiwY** [comment]
I have made it according to your suggestion, it seems to me that it is worse than it was.

![image](https://user-images.githubusercontent.com/24372059/50090777-e0bdb100-0209-11e9-8c1d-169614cd54df.png)

**Comment by jnyrup** [comment]
Have you tried making a fresh console application, that *only* calls `GetWave`?
E.g. try the following and see if memory is still not being released.

```c#
static void Main()
{
    while(true)
    {
        GetWave("path-to-file");
    }
}
```

If you're still seing memory problems, take a snapshot of the memory.
![image](https://user-images.githubusercontent.com/919634/50094356-524e2d00-0213-11e9-80f0-2b50c793ac15.png)
This should give better insights in what objects are not being released.

**Comment by SiiwY** [comment]
![image](https://user-images.githubusercontent.com/24372059/50138710-b3bbdd80-029f-11e9-965f-256bfd7aa64b.png)
It seems to me that the release of memory in this case also works with a delay.

**Comment by jnyrup** [comment]
Try making a snapshot when the memory usage is high, and one once the memory usage is low again.
Then on the second snapshot you should be able to click "diff", to see what objects were released.

**Comment by SiiwY** [comment]
![image](https://user-images.githubusercontent.com/24372059/50221080-52c20180-0394-11e9-90f8-052f8c40326c.png)

![image](https://user-images.githubusercontent.com/24372059/50221144-84d36380-0394-11e9-979b-71fafb5d1942.png)

Is this enough information?

**Comment by jnyrup** [comment]
Looking at those screenshots, I see no sign of a memory leak, only the expected fluctuating memory usage of opening, reading and disposing a grib file.

Can you share the grib file?

Otherwise let's make some better dumps:

```c#
GC.Collect(4,GCCollectionMode.Forced, true);
// take a memory snapshot

for(int i = 0; i < 10000; i++) // or some number that makes GetWave run long enough to expose a potential memory leak
{
    GetWave("path-to-file");
}

GC.Collect(4,GCCollectionMode.Forced, true);

// take a new memory snapshot
```

**Comment by SiiwY** [comment]
ftp://polar.ncep.noaa.gov/history/waves/multi_1/201811/gribs/multi_1.glo_30m.tp.201811.grb2

It will be easier to make the file available to you. Thank you for your huge support.

**Comment by jnyrup** [comment]
After 1500 iterations of the loop the memory usage is "stable" at ~125 mb.
But I explicitly needed to call `GC.Collect()` after each iteration, otherwise I got a memory allocation exception after just 44 iterations, as the memory usage was ~4gb.

I stabilized the memory usage by caching the `Message`s to avoid opening the `GribFile` over and over again.
```c#
List<GribMessage> messages = GetFiles();

List<GribMessage> GetFiles()
{
    using (var file = new GribFile(path))
    {
        return file.ToList();
    }
}

GridCoordinateValue GetWave()
{
    GribMessage msg = messages.Last(_ => true);
    return msg.GridCoordinateValues.Last(_ => true);
}
```

**Comment by SiiwY** [comment]
Thank you very much for the example solution. Unfortunately, your suggestion will not work in my case. I need to read one value from this file and forget about it. Example: I had about 12 hours on the page "latitude = 55 length = 333". I would like to get the wave height from this place. It is not necessary to keep the entire file in memory. :(

**Comment by jnyrup** [comment]
> not necessary to keep the entire file in memory.
not necessary or not possible?

Caching is always a cpu/memory compromise.

**Comment by SiiwY** [comment]
Thank you for your help. Maybe we did not understand a bit. I had a problem that unused objects were not released. What with a large number of requests occupied the entire RAM. I have forced free objects to be released using GC.Collect () ;. And now the application works fine. I wanted to avoid manual release of unused objects.

**Comment by jnyrup** [comment]
I agree that calling `GC.Collect()` isn't optimal and shouldn't be necessary.
I wonder why the garbage collector doesn't seem to kick in earlier.
I'll try to dig more into this.
If I haven't provided any status updates within a month, please poke me.

**Comment by SiiwY** [comment]
Thank you very much for your support! Your help is invaluable. 
I will be waiting for the status update. :)
Best regards, SiiwY

**Comment by jnyrup** [comment]
I found the culprit and a better workaound.

`GribMessage` implements `IDisposable` and `GribFile` implements `IEnumerable<GribMessage>`.
When enumerating a `GribFile` each `GribMessage` should be properly disposed.
When calling e.g. `Linq.First(predicate)` the `GribFile` is enumerated, and `GribMessage`s are created, but only the returned `GribMessage` can be disposed from user-code.

A better workaround that calling `GC.Collect` is to create an overload of `First()` that calls `Dispose` on each non-returned `GribMessage`.

```c#
public static GribMessage First(this GribFile source, Func<GribMessage, bool> predicate)
{
    if (source is null) throw new ArgumentNullException(nameof(source));
    if (predicate is null) throw new ArgumentNullException(nameof(predicate));

    foreach (GribMessage element in source)
    {
        if (predicate(element))
        {
            return element;
        }

        element.Dispose();
    }

    throw new InvalidOperationException("Sequence contains no matching element");
}

private static GridCoordinateValue GetWave(string path, Func<GribMessage, bool> messagePredicate, Func<GridCoordinateValue, bool> gridCoordinateValuePredicate)
{
    using (var file = new GribFile(path))
    {
        using (GribMessage msg = file.First(messagePredicate)) // <-- should resolve to the overload above
        {
            return msg.GridCoordinateValues.Last(gridCoordinateValuePredicate); // <-- regular LINQ
        }
    }
}
```

**Comment by SiiwY** [comment]
Thank you for your time to resolve this matter. I invoke GC and after running the tests the application works correctly. I believe that the information will be useful to other users. Thank you again.
